packages/ui/components/dialog/Dialog.tsx

```diff
@@ -38,15 +50,20 @@ export function Dialog(props: DialogProps) {
         });
         router.push(`${pathname}?${newSearchParams.toString()}`);
       }
-      setOpen(open);
+      setDialogState(open ? DIALOG_STATE.OPEN : DIALOG_STATE.CLOSING);
     };
-    // handles initial state
-    if (!open && newSearchParams.get("dialog") === name) {
```

Comment:
This was quite a tricky handling here. We were modifying `query` earlier(delete operator on it) which is a reference to router.query and thus router.query would have query param dialog removed instantly. In reality, URL would still have that query param. With next/router, it stopped working because `searchParams` is readOnly and creating a new copy of it(i.e. newSearchParams) would loose the reference.

So, I have introduced a life cycle state CLOSING for this case. As soon as Dialog is dismissed, it doesn't go to CLOSED state. It goes to CLOSING state instead and then navigates to CLOSED state when the query param is removed.

Reply:
It fixes this bug: https://www.loom.com/share/a71d1407864343c0bd494e754242c8ab